rust/cubesql/cubesql/src/compile/rewrite/rules/split/functions.rs

```diff
@@ -9,88 +9,52 @@ impl SplitRules {
         &self,
         rules: &mut Vec<Rewrite<LogicalPlanLanguage, LogicalPlanAnalysis>>,
     ) {
-        self.single_arg_pass_through_rules(
-            "trunc",
-            |expr| self.fun_expr("Trunc", vec![expr]),
-            false,
-            rules,
-        );
-        self.single_arg_pass_through_rules(
-            "lower",
-            |expr| self.fun_expr("Lower", vec![expr]),
-            false,
-            rules,
-        );
-        self.single_arg_pass_through_rules(
-            "upper",
-            |expr| self.fun_expr("Upper", vec![expr]),
-            false,
-            rules,
-        );
-        self.single_arg_pass_through_rules(
-            "ceil",
-            |expr| self.fun_expr("Ceil", vec![expr]),
-            false,
-            rules,
-        );
-        self.single_arg_pass_through_rules(
-            "floor",
-            |expr| self.fun_expr("Floor", vec![expr]),
-            false,
-            rules,
-        );
-        self.single_arg_pass_through_rules(
-            "char-length",
-            |expr| self.fun_expr("CharacterLength", vec![expr]),
-            false,
+        // Universal rule to traverse any number of function arguments
+        // PushDown(ScalarFunctionExprArgs(..., arg, ...)) => ScalarFunctionExprArgs(..., PushDown(arg), ...)
+        // ScalarFunctionExprArgs(..., PullUp(arg), ...) => PullUp(ScalarFunctionExprArgs(..., arg, ...))
+        // If your function arguments requires special treatment, avoid generic rewrite like this
+        // PushDown(ScalarFunctionExpr ?fun ?args) => ScalarFunctionExpr ?fun PushDown(?args)
+        // Instead, use direct rewrite like this
+        // PushDown(ScalarFunctionExpr ?fun ScalarFunctionExprArgs(arg1 arg2)) => ScalarFunctionExpr ?fun ScalarFunctionExprArgs(PushDown(arg1) arg2)
+        Self::flat_list_pushdown_pullup_rules(
+            "scalar-fun-var-args",
+            ListType::ScalarFunctionExprArgs,
             rules,
         );
+
+        let fns = [
+            ("Trunc", false),
+            ("Lower", false),
+            ("Upper", false),
+            ("Ceil", false),
+            ("Floor", false),
+            ("CharacterLength", false),
+            ("Substr", false),
+            ("Lpad", false),
+            ("Rpad", false),
+        ];
```

Comment:
Are there any functions with `true` projection on the radars? If no, maybe simplify this to just func names and pass always `false` in loop below?

Reply:
There's `coalesce` and `nullif`, they have projections enabled ATM, but I plan to disable projections for them as well.
In general there's stuff like `to_hex(integer)` or ` chr(integer)` that, I think, can have projections enabled. Are they on the radars? I have no idea :)